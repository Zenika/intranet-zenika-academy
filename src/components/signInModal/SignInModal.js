import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';

import axios from 'axios';


export class SignInModal extends Component {
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      password: '',
      redirectToAdmin: false,
      redirectToUser: false,
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }


  handleChange(e) {
    const { name, value } = e.target;
    this.setState({ [name]: value });
  }

  handleSignIn(user) {
    const { connect, toggleModal } = this.props;
    axios.post('http://localhost:4000/api/users/signin', user)
      .then((res) => {
        sessionStorage.setItem('promoId', `${res.data.promoId}`);
        sessionStorage.setItem('loggedIn', 'true');
        if (res.data.role === 1) {
          this.setState({ redirectToAdmin: true, redirectToUser: false });
        } if (res.data.role === 3 || res.data.role === 2) {
          this.setState({ redirectToUser: true, redirectToAdmin: false });
        }
        return connect();
      })
      .catch((err) => {
        toggleModal(false);
        alert('Impossible de vous authentifier');
        throw new Error(err);
      });
  }

  handleSubmit(e) {
    e.preventDefault();
    const { email, password } = this.state;
    const user = {
      email,
      password,
    };

    this.setState({ email: '', password: '' });
    return this.handleSignIn(user);
  }

  render() {
    const {
      email, password, redirectToUser, redirectToAdmin,
    } = this.state;
    const { handleChange, handleSubmit } = this;
    const { toggleModal } = this.props;
    if (redirectToUser === true) {
      return <Redirect to="/home/user" />;
    }
    if (redirectToAdmin === true) {
      return <Redirect to="/home/admin" />;
    }
    return (
      <section id="modalLogin" className="modal is-active">
        <section className="modal-background" />
        <section className="modal-card">
          <header className="modal-card-head">
            <p className="modal-card-title">Connexion</p>
          </header>
          <section className="modal-card-body">
            <form>
              <section className="field">
                <p className="control has-icons-left">
                  <input
                    id="emailInput"
                    required
                    className="input"
                    type="text"
                    name="email"
                    placeholder="Email"
                    onChange={handleChange}
                    value={email}
                  />
                  <span className="icon is-small is-left" />
                </p>
              </section>
              <section className="field">
                <p className="control has-icons-left">
                  <input
                    id="passwordInput"
                    required
                    className="input"
                    type="password"
                    name="password"
                    placeholder="Password"
                    onChange={handleChange}
                    value={password}
                  />
                  <span className="icon is-small is-left">
                    <i className="fas fa-lock" />
                  </span>
                </p>
              </section>
              <footer className=" modalFooter">
                <button
                  id="submitButton"
                  type="submit"
                  onClick={handleSubmit}
                  className="button is-success"
                >
                  Se connecter
                </button>
                <button
                  id="cancelButton"
                  type="button"
                  onClick={() => toggleModal(false)}
                  className="button"
                >
                  Annuler
                </button>
              </footer>
            </form>
          </section>
        </section>
      </section>
    );
  }
}

export default SignInModal;
