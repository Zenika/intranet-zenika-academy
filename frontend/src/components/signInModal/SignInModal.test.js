import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { SignInModal } from './SignInModal';

Enzyme.configure({ adapter: new Adapter() });

let wrapper;
beforeEach(() => {
  const toggleModal = jest.fn();
  const connect = jest.fn();
  wrapper = shallow(
    <SignInModal toggleModal={toggleModal} connect={connect} />,
  );
});

describe('Modal signIn tests', () => {
  it('Should exist', () => {
    expect(wrapper).toBeDefined();
    expect(wrapper.exists()).toBe(true);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Should have id #modalLogin', () => {
    const container = wrapper.find('#modalLogin');
    expect(container).toHaveLength(1);
  });

  it('Should have a submit button', () => {
    const container = wrapper.find('#submitButton');
    expect(container).toHaveLength(1);
  });

  it('Should change email state with handle change on emailInput', () => {
    const input = wrapper.find('#emailInput');
    const mockEvent = {
      target: {
        name: 'email',
        value: 'test',
      },
    };
    input.simulate('change', mockEvent);
    expect(wrapper.state().email).toEqual('test');
  });

  it('Should change password state with handle change on passwordInput', () => {
    const input = wrapper.find('#passwordInput');
    const mockEvent = {
      target: {
        name: 'password',
        value: 'test',
      },
    };
    input.simulate('change', mockEvent);
    expect(wrapper.state().password).toEqual('test');
  });

  it('Should have cancel button', () => {
    const container = wrapper.find('#cancelButton');
    expect(container).toHaveLength(1);
  });

  it('Should call toggleModal() props on cancel button click', () => {
    const button = wrapper.find('#cancelButton');
    const toggle = wrapper.instance().props.toggleModal;
    button.simulate('click');
    expect(toggle).toHaveBeenCalled();
  });
});
