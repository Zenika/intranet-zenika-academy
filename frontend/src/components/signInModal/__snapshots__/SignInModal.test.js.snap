// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal signIn tests should match the snapshot 1`] = `
<section
  className="modal is-active"
  id="modalLogin"
>
  <section
    className="modal-background"
  />
  <section
    className="modal-card"
  >
    <header
      className="modal-card-head"
    >
      <p
        className="modal-card-title"
      >
        Connexion
      </p>
    </header>
    <section
      className="modal-card-body"
    >
      <form>
        <section
          className="field"
        >
          <p
            className="control has-icons-left"
          >
            <input
              className="input"
              id="emailInput"
              name="email"
              onChange={[Function]}
              placeholder="Email"
              required={true}
              type="email"
              value=""
            />
            <span
              className="icon is-small is-left"
            />
          </p>
        </section>
        <section
          className="field"
        >
          <p
            className="control has-icons-left"
          >
            <input
              className="input"
              id="passwordInput"
              name="password"
              onChange={[Function]}
              placeholder="Password"
              required={true}
              type="password"
              value=""
            />
            <span
              className="icon is-small is-left"
            >
              <i
                className="fas fa-lock"
              />
            </span>
          </p>
        </section>
        <footer
          className=" modalFooter"
        >
          <button
            className="button is-success"
            id="submitButton"
            onClick={[Function]}
            type="submit"
          >
            Se connecter
          </button>
          <button
            className="button"
            id="cancelButton"
            onClick={[Function]}
            type="button"
          >
            Annuler
          </button>
        </footer>
      </form>
    </section>
  </section>
</section>
`;
